### PR TITLE
⏩ Train on completion only

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LeonEricsson** [comment]
Suggestion to update the [DataCollator example](https://github.com/huggingface/trl/blob/3a2788b47ec542d163bb180a6fe461a341dceb7d/trl/trainer/sft_trainer.py#L87) to:

```
    Examples:
    ```python
    >>> from trl import DataCollatorForLanguageModeling
    >>> collator = DataCollatorForLanguageModeling(pad_token_id=0)
    >>> examples = [
    ...     {"input_ids": [1, 2, 3, 4], "completion_mask": [0, 0, 1, 1]},
    ...     {"input_ids": [5, 6, 7], "completion_mask": [0, 1, 1]}
    ... ]
    >>> collator(examples)
    {'input_ids': tensor([[1, 2, 3, 4],
                          [5, 6, 7, 0]]),
     'attention_mask': tensor([[1, 1, 1, 1],
                               [1, 1, 1, 0]]),
     'labels': tensor([[-100, -100,    3,    4],
                       [ -100,    6,    7, -100]])}
    ```
```

given that `completion_only_loss` is true by default

File: tests/test_sft_trainer.py [hunk 1]
```diff
@@ -1064,23 +1043,6 @@ def test_train_model_torch_dtype(self):
...
-            with self.assertRaises(ValueError) as context:
-                SFTTrainer(
-                    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5", args=training_args, train_dataset=dataset
-                )
-            self.assertIn(
-                "Invalid `torch_dtype` passed to `SFTConfig`. Expected either 'auto' or a string representing "
-                "a `torch.dtype` (e.g., 'float32'), but got -1.",
-                str(context.exception),
-            )
-
```
Comment (qgallouedec) [inline tests/test_sft_trainer.py:1083]: This is not related to the core change of this PR.
With the new serialisation logic of `TrainingArguments`, passing a wrong dtype fails when you instantiate the `TrainingArguments`. There is no need for such test anymore

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -500,16 +514,6 @@ def _func(example):
...
-            # If the dataset is prompt-completion, convert it to language modeling type
-            first_example = next(iter(dataset))
-            if "prompt" in first_example.keys() and "completion" in first_example.keys():
-                key = "messages" if is_conversational(first_example) else "text"
-
-                def concat_prompt_completion(example):
-                    return {key: example["prompt"] + example["completion"]}
-
-                dataset = dataset.map(concat_prompt_completion, remove_columns=["prompt", "completion"])
-
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:512]: This concatenation needs to be removed, as we loses the information about where the completion starts. This completion is now managed in `tokenize`.

**Comment by qgallouedec** [comment]
> Suggestion to update the [DataCollator example](https://github.com/huggingface/trl/blob/3a2788b47ec542d163bb180a6fe461a341dceb7d/trl/trainer/sft_trainer.py#L87) to:

Thanks! Done in

Done in 7f7f2a4107bf60c6e8e5aafe9c99852f70c8bef9

**Review by shirinyamani — APPROVED** [review]
